### DISCUSSION TITLE
[Angular, NX] Incomplete documentation in the Args table (new behavior?)

### DISCUSSION THREAD

**Discussion by KyneSilverhide** [storybookjs/storybook#28783] [Help] [answered]
### Summary

Hi,

We recently migrated our storybook to a monorepo. While the basic features do work, we've found out that the argument tables is now mostly empty.
BEFORE the migration, the whole table was filled with all "possible" properties from the component

![2024-08-02 10_15_18-Conversation _ Nicolas Demerrisse _ Microsoft Teams](https://github.com/user-attachments/assets/666a56fe-6c19-4942-a223-7370ebc573d3)

And now, it's "empty".
![2024-08-02 10_15_29-Conversation _ Nicolas Demerrisse _ Microsoft Teams](https://github.com/user-attachments/assets/e73ea101-a051-4c3b-b16a-e3e87b15ba14)

If I look at the code, the arguments are indeed empty but.. the code is the same as before
`
export const Example: Story = {
  args: {
  }
}`

If I put a basic arg, it will be shown in the table, but everything else is still missing.
My concern is... how did this behavior change? **Before, all properties (Inputs, etc) from the components were always displayed. And now, only the args from the FIRST example are displayed in the table..;**

### Additional information

**package.json**
```
  "@storybook/addon-a11y": "^7.6.16",
      "@storybook/addon-docs": "^7.4.2",
      "@storybook/addon-essentials": "7.6.20",
      "@storybook/addon-interactions": "^7.5.3",
      "@storybook/addon-links": "^7.4.2",
      "@storybook/angular": "7.6.20",
      "@storybook/core-server": "7.6.20",
      "@storybook/jest": "^0.2.3",
      "@storybook/test-runner": "^0.13.0",
      "@storybook/testing-library": "^0.2.2",
      "@storybook/theming": "^7.4.2",
  ...
      "nx": "18.2.4",
  ...
      "@angular/core": "17.3.12",
```

**main.ts**

```
import type { StorybookConfig } from '@storybook/angular';

const config: StorybookConfig = {
  stories: [
    '../src/lib/**/*.stories.@(js|jsx|ts|tsx|mdx)',
    '../src/lib/components/**/*.mdx'
  ],
  addons: [
    '@storybook/addon-links',
    '@storybook/addon-essentials',
    '@storybook/addon-interactions',
    '@storybook/addon-a11y'
  ],
  framework: {
    name: '@storybook/angular',
    options: {},
  },
  core: {
    builder: "@storybook/builder-webpack5"
  },
  docs: {
    autodocs: 'tag',
  },
  staticDirs: ['./public']
};

export default config;

// To customize your webpack configuration you can use the webpackFinal field.
// Check https://storybook.js.org/docs/react/builders/webpack#extending-storybooks-webpack-config
// and https://nx.dev/recipes/storybook/custom-builder-configs
```

**Comment by Marklb** [comment]
- What do you have for `"compodoc"` and `"compodocArgs"` in your `angular.json`(or Nx equivalent) builder for storybook.
- What is your compodoc version?
- Is `setCompodocJson` called anywhere? It is in `.storybook/preview.ts` by default.

Without more information, I am going to assume it is the config, but there isn't enough information to be sure. 

Some things to check:
- Does a `documentation.json` file get generated, when you build Storybook?
  - If not, compodoc isn't configured. Add `"compodoc": true, "compodocArgs": [ "-e", "json", "-d", "." ],` to your builder, in `angular.json`(or Nx equivalent).
- Does `documentation.json` contain the properties for the component class?
  - ~~If not, add `import docJson from "../documentation.json"; setCompodocJson(docJson);` to `.storybook/preview.ts`.~~
  - If it does contain the properties then I would assume your import is missing or wrong. Add `import docJson from "../documentation.json"; setCompodocJson(docJson);` to `.storybook/preview.ts`. If the import is already there, make sure the import path is correct.

**Reply by KyneSilverhide** [reply]
I can't answer everything for now, because I'm on holiday and I don't have access to the code anymore.

BUT, yes, a documentation.json was indeed created each time. I deleted it to be sure, and it was updated each build. And at the proper destination I set (instead of the root folder by default). 
Also, If I add any story or example, I do see the changes. 

So yes, we do call `setCompodocJson` in `.storybook/preview.ts`
And yes, I also added the `compodoc` and `compodocArgs` to the builder. 

So the concern is really regarding the arguments...

**Reply by KyneSilverhide** [reply]
Ok, here are all the answers :

* What do you have for "compodoc" and "compodocArgs" in your angular.json(or Nx equivalent) builder for storybook.
 ```
"compodoc": true,
"compodocArgs": ["-e", "json", "-d", "libs/care-common-lib"]
```

* What is your compodoc version?
`1.1.25`

* Is setCompodocJson called anywhere? It is in .storybook/preview.ts by default.
Yes, it is
```
import type {Preview} from '@storybook/angular';
import {setCompodocJson} from '@storybook/addon-docs/angular';
import docJson from "../documentation.json";
setCompodocJson(docJson);

const preview: Preview = {
...
```

* Does a documentation.json file get generated, when you build Storybook? 
Yes, it's re-generated each time.

* Does documentation.json contain the properties for the component class? 
This is already added (see snippet in previous question)

**Reply by Marklb** [reply]
Since you have `"compodocArgs": ["-e", "json", "-d", "libs/care-common-lib"]`, I will assume your `preview.ts` is at `libs/care-common-lib/.storybook/preview.ts`. Is that correct?

> - Does documentation.json contain the properties for the component class?
> This is already added (see snippet in previous question)

If you mean the one showing the import then that isn't what I am asking, but I do see that I made the question confusing. I don't know why I said to add the import if not. I meant can you confirm that the content of `documentation.json` has something like the following and if it does then make sure the import is correct.
```json
{
  ...
  "components": [
    {
      "name": "ExampleComponent",
      "id": "component-ExampleComponent-f1870f55718477db5...",
      "file": "projects/my-project/src/lib/example.component.ts",
      "encapsulation": [],
      "entryComponents": [],
      "inputs": [],
      "outputs": [],
      "providers": [],
      "selector": "lib-example",
      "styleUrls": [
        "./example.component.scss"
      ],
      "styles": [],
      "templateUrl": [
        "./example.component.html"
      ],
      "viewProviders": [],
      "hostDirectives": [],
      "inputsClass": [
        {
          "name": "title",
          "deprecated": false,
          "deprecationMessage": "",
          "line": 19,
          "type": "string | undefined | null",
          "decorators": []
        }
      ],
      "outputsClass": [],
      "propertiesClass": [],
      "methodsClass": [],
      "deprecated": false,
      "deprecationMessage": "",
      "hostBindings": [],
      "hostListeners": [],
      "standalone": false,
      "imports": [],
      "description": "",
      "rawdescription": "\n",
      "type": "component",
      "sourceCode": "...",
      "assetsDirs": [],
      "styleUrlsData": [],
      "stylesData": "",
      "extends": [],
      "templateData": "..."
    },
  ],
  ...
}
```

**Reply by KyneSilverhide** [reply]
Yes, exactly.
![image](https://github.com/user-attachments/assets/0e56b75e-06eb-4e9f-87b9-b47fd1a43847)


And indeed, the documentation.json is contianing that kind of content. It's also properly formatted.
![image](https://github.com/user-attachments/assets/32d922cc-a89f-405e-ab83-1366be3ee2e8)
![image](https://github.com/user-attachments/assets/59a62718-01c2-4ea6-a617-2d4f61b99db7)

Regarding the properties, I'll check the JSON because it's a bit large...

**Reply by KyneSilverhide** [reply]
Something is strange, or maybe I'm not understanding what should be generated inside documentation.json... but I can't find this specific component.

I do see it on the menu
![image](https://github.com/user-attachments/assets/5885c92f-5420-491a-b54c-498b9ee9ca39)

And there is content (well, the stories are displayed, and any change I make in the code is visible on the web page...) :
![image](https://github.com/user-attachments/assets/ded5452a-e6ac-4750-abfc-56aa2069dee7)

But I don't see anything related to this component in documentation.json...
```
const meta: Meta<CareDatePickerComponent> = {
  title: 'Design System/Components/Forms/DatePicker',
  component: CareDatePickerComponent,
  tags: ['autodocs'],
  parameters: {
    docs: {
      description: {
        component: StatusEnum.COMPLETE
      },
    },
  },
```

![image](https://github.com/user-attachments/assets/ceb368ae-62b6-42cb-970d-7035ac9009bb)

**Reply by Marklb** [reply]
For the components that do show up in `documentation.json`, do they not show the ArgsTable or do those work?

For the ones that do not show, it may be your tsconfig. compodoc doesn't have a feature that allows it to integrate with other tools and update independent files on change, etc. It is run independently, before storybook is built. So, rendering in storybook and not showing up in `documentation.json` are not really related.

If you wanted to just test compodoc independently, you could run `compodoc -p ./tsconfig.json -e json -d .`. When the builder options have `"compodoc": true`, storybook should be passing the tsconfig defined in your builder options (`angular.json`/`workspace.json`).

For my projects, I use a tsconfig specific to storybook that extends my library tsconfig. The reason I do that is because I had ts files that were not included directly and compodoc wasn't including them. *I can't remember the specific reason, but I will update this if I remember.* Ideally, you shouldn't need to do this, but just to start narrowing down where the problem is you could try adding a tsconfig like the following. Then set the tsconfig in your storybook builder options to it.
```json
// tsconfig.sb.json
{
  "extends": "./tsconfig.lib.json",
  "include": [
    "**/*.stories.ts",
    "**/*.ts"
  ],
  "exclude": [
    "**/*.spec.ts"
  ],
}
```

**Reply by KyneSilverhide** [reply]
I've not found much components that work. This one seems to have an arguments table:
![image](https://github.com/user-attachments/assets/13aaccda-7de0-46c5-9082-8d3622054ca1)

But I can't find it in documentation.json...  so yeah, it's not really linked.

We also use a specific tsconfig.json, but we have so many "extends" that I'm not even sure which one is the proper one... 
But we do have one in the lib, and one in storybook extending the one in the lib.

Edit : so here is the tsconfig.json we use : 
```
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "types": [
      "node"
    ],
    "allowSyntheticDefaultImports": true,
    "resolveJsonModule": true
  },
  "exclude": [
    "../src/test.ts",
    "../src/**/*.spec.ts",
    "../src/lib/components/**/*.spec.ts"
  ],
  "include": [
    "../src/**/*.stories.ts",
    "./preview.ts"
  ]
}

```

**Reply by Marklb** [reply]
With that tsconfig, I would expect all the components imported by a stories file to be parsed by compodoc, but I haven't looked at how tsconfig include and compodoc work in a while. If you aren't seeing the components in the `documentation.json` then it seems like the problem has been narrowed down to compodoc, at least.

For the one that has an ArgsTable, all the controls, except maybe `styleClass`, have a value. So, could it be generating those from `args`/`argTypes` manually set for that story?

I would start running compodoc independently, adding specific files to the tsconfig `"include"` or `"files"` to see if it at least parses them. Then figure out what could cause your current tsconfig to not be finding them.

**Reply by KyneSilverhide** [reply]
i tried to just run compodoc, both from the nx root and from the lib folder, targeting the proper tsconfig.json (instead of the parent, that would result in a much larger documentation.json including all code...)

`compodoc -p libs/care-common-lib/.storybook/tsconfig.json -e json -d .`

But all I get is a compodoc file with no "component". Only "miscellanous". 
![image](https://github.com/user-attachments/assets/7e3082e7-8f07-4fe2-aa02-6a07a45df327)

And there, i do find the stories about one of the "broken" component.
```
{
                "name": "Example",
                "ctype": "miscellaneous",
                "subtype": "variable",
                "file": "libs/care-common-lib/src/lib/components/care-date-picker/care-date-picker.stories.ts",
                "deprecated": false,
                "deprecationMessage": "",
                "type": "Story",
                "defaultValue": "{\r\n  args: {\r\n  }\r\n}"
            },
```


**For the one that has an ArgsTable, all the controls, except maybe styleClass, have a value. So, could it be generating those from args/argTypes manually set for that story?**
Well, more or less. it is only using the args explicitely set in the story from the FIRST story. So if the first sotry is the basic one with no config... the table is empty. And all other tables seems to be empty too...

But before (as seen of the screenshot from the initial post), it somehow auto-generated all possible values in a table...
And the code from the stories have not changed since.

**Reply by KyneSilverhide** [reply]
We have just migrated to storybook 8, but the problem remains.

I think it's somehow related to "autodocs" not working properly... but I don't know why.
![image](https://github.com/user-attachments/assets/3b722788-054e-4585-be03-0f6f08dd721c)

I've also explicitely added autodocs to one of the broken components, but to no avail...
![image](https://github.com/user-attachments/assets/72ee4e8d-e94d-4311-a74a-84415ed01849)

**Reply by Marklb** [reply, marked as answer]
"autodocs" is not necessarily related to the ArgTable. It basically just determines if the documentation pages should automatically be generated. In the Angular framework, the argTypes are generated purely from the json file that compodoc outputs, unless the story manually defines them.

The stories files are not important in the compodoc output. The tsconfig that compodoc receives only has to be able to find your component files.

If you made a project that reproduces your problem that I could clone then I could probably find the issue fairly quickly.

To break this down to a very simple example, just to confirm that the problem is your tsconfig, if my assumption is correct. The following steps are something I am doing myself, so I don't overlook something.

I am going to assume you have a component at `src/app/example.component.ts`.

Then I am going to create a simple tsconfig `tsconfig.example.json`, which does nothing except specify a path directly to the component file. So, I know there is no reason for compodoc to not find that file.

```json
{
  "files": [
    "src/app/example.component.ts"
  ]
}
```

Now, I will run `compodoc -p tsconfig.example.json -e json -d .`.

To eliminate any path resolving issues to the compodoc output, I will copy the contents of `documentation.json` and paste it directly in `.storybook/preview.ts`.

```ts
import type { Preview } from "@storybook/angular";
import { setCompodocJson } from "@storybook/addon-docs/angular";
// import docJson from "../documentation.json";
setCompodocJson({
  "pipes": [],
  "interfaces": [],
  "injectables": [],
  "guards": [],
  "interceptors": [],
  "classes": [],
  "directives": [],
  "components": [
      {
          "name": "ExampleComponent",
          "id": "component-ExampleComponent-f983e02c1dd4624ac7d07dd98cbe611c8401fe7fcb395e58a93e6878c6ebd75f2511d39cf42724ccde2047db98406e335c58655b50fc5ac619d8ba5f687a934e",
          "file": "src/app/example.component.ts",
          "encapsulation": [],
          "entryComponents": [],
          "inputs": [],
          "outputs": [],
          "providers": [],
          "selector": "example",
          "styleUrls": [],
          "styles": [],
          "template": "",
          "templateUrl": [],
          "viewProviders": [],
          "hostDirectives": [],
          "inputsClass": [
              {
                  "name": "input1",
                  "defaultValue": "''",
                  "deprecated": false,
                  "deprecationMessage": "",
                  "line": 11,
                  "type": "string",
                  "decorators": []
              }
          ],
          "outputsClass": [],
          "propertiesClass": [],
          "methodsClass": [],
          "deprecated": false,
          "deprecationMessage": "",
          "hostBindings": [],
          "hostListeners": [],
          "standalone": true,
          "imports": [],
          "description": "",
          "rawdescription": "\n",
          "type": "component",
          "sourceCode": "import { Component, Input } from '@angular/core';\r\n\r\n@Component({\r\n  selector: 'example',\r\n  standalone: true,\r\n  imports: [],\r\n  template: ``,\r\n  styles: [],\r\n})\r\nexport class ExampleComponent {\r\n  @Input() input1: string = ''\r\n}\r\n",
          "assetsDirs": [],
          "styleUrlsData": "",
          "stylesData": "",
          "extends": []
      }
  ],
  "modules": [],
  "miscellaneous": [],
  "routes": [],
  "coverage": {
      "count": 0,
      "status": "low",
      "files": [
          {
              "filePath": "src/app/example.component.ts",
              "type": "component",
              "linktype": "component",
              "name": "ExampleComponent",
              "coveragePercent": 0,
              "coverageCount": "0/2",
              "status": "low"
          }
      ]
  }
});

const preview: Preview = {
  parameters: {
    controls: {
      matchers: {
        color: /(background|color)$/i,
        date: /Date$/i,
      },
    },
  },
};

export default preview;
```

**Reply by KyneSilverhide** [reply]
Ok, now that I know how this args table is built, I was able to investigate a bit more.

As you expected, issue was in the tsconfig.json. During the NX migration,  a new tsconfig.json was created... and was used instead of the old one. 
I've merged them and now I can finally see the args.
![image](https://github.com/user-attachments/assets/92873245-32d2-4007-9601-29e93f1bb386)


But there's still an issue... whenever I'm updating anything, I've got an error..
![image](https://github.com/user-attachments/assets/4d3f48b1-1fd4-44b5-a9ed-c903d84f0ebf)
`Error: NG05104: The selector "story--design-system-components-forms-datepicker--example--primary-inner[story--design-system-components-forms-datepicker--example--primary-inner-4]" did not match any elements. Find more at https://angular.io/errors/NG05104`

Maybe I should create anther discussion for this one...

**Reply by KyneSilverhide** [reply]
Edit : ok sorry, this was my fault... we used specific arguments before that I forgot to add back with all my tests..

`"compodocArgs": ["-e", "json", "--disablePrivate", "--disableProtected", "--disableInternal", "--disableLifeCycleHooks", "-d", "libs/care-common-lib/.storybook"],
`

It's fine now ;)

**Comment by Walnussbaer** [comment]
I am facing the same issue, but in a Vue project :( (https://github.com/storybookjs/storybook/discussions/28817)

**Reply by Marklb** [reply]
Just for reference, we resolved the Vue issue and it was due to incorrect vite plugin order. Angular's builder is different and doesn't generate documentation data that way, so it is unrelated.
